File: sidecar/src/protocol.ts
/**
 * NDJSON wire protocol between the Rust host and this Node sidecar.
 *
 * One JSON object per line. stdout carries protocol frames only; all
 * diagnostics go to stderr so a stray log can never corrupt the stream.
 */

import type { PermissionMode, PermissionUpdate } from '@anthropic-ai/claude-agent-sdk';

/* ------------------------------------------------------------------ */
/* Host -> Sidecar                                                     */
/* ------------------------------------------------------------------ */

export type HostFrame =
  | { t: 'start'; id: string; cfg: StartConfig }
  | { t: 'send'; id: string; text: string }
  | { t: 'interrupt'; id: string }
  | { t: 'set_mode'; id: string; mode: PermissionMode }
  | { t: 'set_model'; id: string; model?: string }
  | { t: 'permission_reply'; id: string; reqId: string; result: PermissionReply }
  | { t: 'ask'; id: string; reqId: string; what: AskKind }
  | { t: 'close'; id: string }
  | { t: 'shutdown' };

/** Introspection the UI can pull on demand. */
export type AskKind = 'context' | 'mcp_status' | 'commands' | 'models';

export interface StartConfig {
  /** Working directory. Decides which CLAUDE.md, .claude/ and .mcp.json load. */
  cwd: string;
  /** First user turn. Omit to open an idle session. */
  prompt?: string;
  /** Prior SDK session id to continue. */
  resume?: string;
  /** Continue `resume` as a new branch instead of extending it. */
  forkSession?: boolean;
  model?: string;
  permissionMode?: PermissionMode;
  /**
   * Full environment for the agent process, resolved by the host from a
   * login shell. Without this a GUI-launched app hands the agent a stub
   * PATH and every npx-based MCP server fails to start.
   */
  env?: Record<string, string>;
  /** Explicit `claude` binary path, when the host found one off PATH. */
  claudePath?: string;
}

export type PermissionReply =
  | { behavior: 'allow'; updatedInput?: Record<string, unknown>; updatedPermissions?: PermissionUpdate[] }
  | { behavior: 'deny'; message: string; interrupt?: boolean };

/* ------------------------------------------------------------------ */
/* Sidecar -> Host                                                     */
/* ------------------------------------------------------------------ */

export type SidecarFrame =
  | { t: 'event'; id: string; ev: AgentEvent }
  | { t: 'permission_request'; id: string; reqId: string; req: PermissionRequest }
  | { t: 'permission_settled'; id: string; reqId: string }
  | { t: 'reply'; reqId: string; ok: true; data: unknown }
  | { t: 'reply'; reqId: string; ok: false; error: string }
  | { t: 'ready'; pid: number };

export interface PermissionRequest {
  toolName: string;
  input: Record<string, unknown>;
  /** Prompt sentence rendered by the SDK, e.g. "Claude wants to read foo.txt". */
  title?: string;
  /** Short noun phrase for a button label, e.g. "Read file". */
  displayName?: string;
  /** Why the request fired at all. */
  decisionReason?: string;
  /** Path that tripped a directory boundary, when applicable. */
  blockedPath?: string;
  /**
   * Rules to persist if the user picks "always allow". Hand these straight
   * back as `updatedPermissions`.
   */
  suggestions?: PermissionUpdate[];
}

/**
 * Normalized event stream. Every executor (Agent SDK today, other agent
 * CLIs later) emits this shape so one UI renders all of them.
 */
export type AgentEvent =
  | {
      kind: 'init';
      sessionId: string;
      model: string;
      cwd: string;
      permissionMode: string;
      tools: string[];
      /** Terminal-parity evidence: what actually loaded for this session. */
      mcpServers: Array<{ name: string; status: string }>;
      slashCommands: string[];
      skills: string[];
      plugins: unknown[];
      pluginErrors: unknown[];
      mcpServerErrors: unknown[];
    }
  | { kind: 'text_delta'; text: string }
  | { kind: 'text'; text: string }
  | { kind: 'thinking'; text: string }
  | { kind: 'tool_call'; toolUseId: string; name: string; input: unknown }
  | { kind: 'tool_result'; toolUseId: string; content: unknown; isError: boolean }
  | { kind: 'turn_end' }
  | {
      kind: 'done';
      subtype: string;
      isError: boolean;
      costUsd?: number;
      durationMs?: number;
      numTurns?: number;
      result?: string;
      sessionId?: string;
    }
  | { kind: 'error'; message: string }
  | { kind: 'closed' };

File: src-tauri/parked/protocol.rs
//! Rust side of the sidecar NDJSON protocol. Mirrors `sidecar/src/protocol.ts`.
//!
//! Frames the core acts on are typed. Frames it only forwards to the UI keep
//! their payload as raw JSON, so the sidecar and the frontend can evolve the
//! event shape without a Rust change in the middle.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/* ------------------------------------------------------------------ */
/* Host -> Sidecar                                                     */
/* ------------------------------------------------------------------ */

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "t")]
pub enum HostFrame {
    #[serde(rename = "start")]
    Start { id: String, cfg: StartConfig },
    #[serde(rename = "send")]
    Send { id: String, text: String },
    #[serde(rename = "interrupt")]
    Interrupt { id: String },
    #[serde(rename = "set_mode")]
    SetMode { id: String, mode: String },
    #[serde(rename = "set_model")]
    SetModel {
        id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        model: Option<String>,
    },
    #[serde(rename = "permission_reply")]
    PermissionReply {
        id: String,
        #[serde(rename = "reqId")]
        req_id: String,
        result: PermissionReply,
    },
    #[serde(rename = "ask")]
    Ask {
        id: String,
        #[serde(rename = "reqId")]
        req_id: String,
        what: String,
    },
    #[serde(rename = "close")]
    Close { id: String },
    #[serde(rename = "shutdown")]
    Shutdown,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct StartConfig {
    pub cwd: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resume: Option<String>,
    #[serde(rename = "forkSession", skip_serializing_if = "Option::is_none")]
    pub fork_session: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(rename = "permissionMode", skip_serializing_if = "Option::is_none")]
    pub permission_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub env: Option<HashMap<String, String>>,
    #[serde(rename = "claudePath", skip_serializing_if = "Option::is_none")]
    pub claude_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "behavior")]
pub enum PermissionReply {
    #[serde(rename = "allow")]
    Allow {
        #[serde(rename = "updatedInput", skip_serializing_if = "Option::is_none")]
        updated_input: Option<serde_json::Value>,
        #[serde(
            rename = "updatedPermissions",
            skip_serializing_if = "Option::is_none"
        )]
        updated_permissions: Option<serde_json::Value>,
    },
    #[serde(rename = "deny")]
    Deny {
        message: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        interrupt: Option<bool>,
    },
}

impl PermissionReply {
    pub fn allow() -> Self {
        Self::Allow {
            updated_input: None,
            updated_permissions: None,
        }
    }

    pub fn deny(message: impl Into<String>) -> Self {
        Self::Deny {
            message: message.into(),
            interrupt: None,
        }
    }
}

/* ------------------------------------------------------------------ */
/* Sidecar -> Host                                                     */
/* ------------------------------------------------------------------ */

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "t")]
pub enum SidecarFrame {
    #[serde(rename = "ready")]
    Ready { pid: u32 },
    #[serde(rename = "event")]
    Event {
        id: String,
        ev: serde_json::Value,
    },
    #[serde(rename = "permission_request")]
    PermissionRequest {
        id: String,
        #[serde(rename = "reqId")]
        req_id: String,
        req: PermissionRequest,
    },
    #[serde(rename = "permission_settled")]
    PermissionSettled {
        id: String,
        #[serde(rename = "reqId")]
        req_id: String,
    },
    #[serde(rename = "reply")]
    Reply {
        #[serde(rename = "reqId")]
        req_id: String,
        ok: bool,
        #[serde(default)]
        data: serde_json::Value,
        #[serde(default)]
        error: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionRequest {
    #[serde(rename = "toolName")]
    pub tool_name: String,
    pub input: serde_json::Value,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default, rename = "displayName")]
    pub display_name: Option<String>,
    #[serde(default, rename = "decisionReason")]
    pub decision_reason: Option<String>,
    #[serde(default, rename = "blockedPath")]
    pub blocked_path: Option<String>,
    /// Rules to persist when the user picks "always allow". Opaque here —
    /// they are handed straight back to the SDK.
    #[serde(default)]
    pub suggestions: Option<serde_json::Value>,
}

impl PermissionRequest {
    /// The Bash command line, when this is a Bash request. Used by the rule
    /// engine and for the UI's compact one-line summary.
    pub fn bash_command(&self) -> Option<&str> {
        if self.tool_name != "Bash" {
            return None;
        }
        self.input.get("command")?.as_str()
    }
}
